raiden/channel/netting_channel.py

```diff
@@ -205,6 +205,8 @@ def __init__(
         self.received_transfers = list()
         self.sent_transfers = list()
 
+        self.identifier_to_directtransfer_amount = dict()
```

Comment:
This attribute needs to be saved with the raiden snapshot and restored.

Also, the REST api is exposing a single interface for direct *and* mediated transfers, so the identifiers for both of these transfers need to be shared.

Reply:
I'm not sure we should save it.. in some time, we could end up with too much used identifiers, the client interface would need to know which `identifiers` were already used to avoid them. I think this `identifier` logic implemented in this PR only makes sense in a somehow short timeframe, as the session lifespan.
About the mediated transfers, I think although they use the same attribute (`identifier`), they have its own logic for handling it (locks, etc), I'm not sure this idempotency logic could be applied to mediated transfers in the same way.

Reply:
> in some time, we could end up with too much used identifiers
What is too much? Why is this a problem? A hashtable is O(1) so there is no performance penalty and we can use the sqlite database if memory becomes an issue.

> the client interface would need to know which identifiers were already used to avoid them.
Why? It can use a random identifier and retry if it was already used. Identifiers don't need to be sequential. Also, don't assume there is only one client using raiden.

>  I think this identifier logic implemented in this PR only makes sense in a somehow short timeframe, as the session lifespan.
We don't know how short/long a session is, this has no value, if a user restarts raiden quickly then the REST request would not be idempotent.

> I'm not sure this idempotency logic could be applied to mediated transfers in the same way.
I'm not saying the same logic should be used (nor I'm saying it shouldn't), what I'm saying is that the *identifiers are shared*.

Reply:
> Why? It can use a random identifier and retry if it was already used. Identifiers don't need to be sequential. Also, don't assume there is only one client using raiden.

I think this catch-and-retry pattern isn't good. But, of course, it's client's problem. I'm planning to use a time-based identifier generation (`timestamp + randint(0, 1000)`) anyway, so identifier colision shouldn't be an issue, even with multiple clients.

> if a user restarts raiden quickly then the REST request would not be idempotent

I don't think we should care for this case. If Raiden was restarted, we can assume clients were too (webUI), or at least, are aware of it, and wouldn't expect a new transfer call to be idempotent (i.e.: the new call would act as usual, as a new transfer request). Remember one of the main issues this idempotency is aiming to solve is transfers being repeated when the socket for the REST call got stuck, which wouldn't happen on restart, as all sockets connection would be closed.
Anyway, I'm not sure `DirectTransfer` can be easily [de]serialized. I'll try to check if it can be done easily, and add this information to `Channel` serialization.

> the identifiers are shared

Yes, I know, but if the following (or previous) codepath is different, what's the matter? I think we can care (and provide idempotency) only for direct transfers, but I'm not sure. What do you think, @LefterisJP ?

Reply:
So my 3 cents on this:

1. If we save identifiers we save them for all transfers, irrespective of them being direct or mediated or else this change will make no sense.

2. The catch and retry pattern can be fine, and as you said it's a client side issue.

3. Regarding serialization and deserialization of the identifier:

> We don't know how short/long a session is, this has no value, if a user restarts raiden quickly then the REST request would not be idempotent.
> I don't think we should care for this case. If Raiden was restarted, we can assume clients were too (webUI), or at least, are aware of it, and wouldn't expect a new transfer call to be idempotent (i.e.: the new call would act as usual, as a new transfer request).

So if I understand correctly you are proposing that we should not be serialiizing the mapping because the idempotency you want to achieve would be session only and not persist through raiden restarts?

I can see where you are coming from this but it seems hacky. And our current serialization and deserialization itself is a hack that "Should go away soon"™. I am really reluctant to have us put a hack on top of another hack.